src/lib/homestore_backend/heap_chunk_selector.cpp

```diff
@@ -130,11 +135,44 @@ csharedChunk HeapChunkSelector::select_specific_chunk(const chunk_num_t chunkID)
     if (vchunk.get_internal_chunk()) {
         auto& avalableBlkCounter = it->second->available_blk_count;
         avalableBlkCounter.fetch_sub(vchunk.available_blks());
+        remove_chunk_from_defrag_heap(vchunk.get_chunk_id());
     }
 
     return vchunk.get_internal_chunk();
 }
 
+// select_chunk_for_GC will only be called when GC is triggered, and will return the chunk with the most
+// defrag blocks
+csharedChunk HeapChunkSelector::select_chunk_for_GC() {
+    chunk_num_t chunkID{0};
+    // the chunk might be seleted for creating shard. if this happens, we need to select another chunk
+    for (;;) {
+        {
+            std::lock_guard< std::mutex > lg(m_defrag_mtx);
+            if (m_defrag_heap.empty()) break;
+            chunkID = m_defrag_heap.top().get_chunk_id();
+        }
+        auto chunk = select_specific_chunk(chunkID);
+        if (chunk) return chunk;
+    }
+    return nullptr;
+}
+
+void HeapChunkSelector::remove_chunk_from_defrag_heap(const chunk_num_t chunkID) {
+    std::vector< VChunk > chunks;
+    chunks.reserve(m_defrag_heap.size());
```

Comment:
size() should be called when lock is hold, right?  The rest part is fine to me.

Reply:
ack